keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -3,703 +3,740 @@
+Object storage has been rising in popularity in the recent years as an alternative form of storage to filesystems and block devices. Object storage paradigm promotes disaggregation of compute and storage. This is done by making data available over the network, rather than locally. Disaggregated architectures allow compute workloads to be stateless, which consequently makes them easier to manage, scale and automate. 
 
 ## User Stories
 
-### Admin
+We define 3 kinds of stakeholders:
+
++ **Admins** - Members with escalated privileges and authority to manage site wide/cluster wide policies to control access, assign storage quotas and other concerns related to hardware or application resources.
++ **Application Developer/Operator (User)** - Members with privileges to run workloads, request storage for them, and manage resources inside a specific namespace.
++ **Object Storage Provider (OSP)** - Vendors offering Object Storage capabilities
+
+#### Admin
 
-- As a cluster administrator, I can control access to new and existing buckets when accessed from the cluster, regardless of the backing object store.
++ Establish cluster/site wide policies for data redundancy, durability and other data related parameters
++ Establish cluster/site wide policies for access control over buckets
++ Enable/restrict users who can create, delete and reuse buckets
 
-### User
+#### DevOps
 
-- As a developer, I can define my object storage needs in the same manifest as my workload, so that deployments are streamlined and encapsulated within the Kubernetes interface.
-- As a developer, I can define a manifest containing my workload and object storage configuration once, so that my app may be ported between clusters as long as the storage provided supports my designated data path protocol.
-- As a developer, I want to create a workload controller which is bucket API aware, so that it can dynamically connect workloads to underlying object storage.
++ Request a bucket to be created and/or provisioned for workloads
++ Request access to be created/deleted for a specific bucket
++ Request deletion of a created bucket
+
+### Object Storage Provider (OSP)
+
++ Integrate with the official standard for orchestrating Object Storage resources in Kubernetes
 
 ## Goals
 
-+ Specify object storage Kubernetes APIs for the purpose of orchestrating/managing object store operations.
-+ Implement a Kubernetes controller automation design with support for pluggable provisioners.
-+ Support workload portability across clusters.
-+ As MVP, be accessible to the largest groups of consumers by supporting the major object storage protocols (S3, Google Cloud Storage, Azure Blob) while being extensible for future protocol additions..
-+ Present similar workflows for both greenfield and brownfield bucket operations.
+#### Functionality
 
-## Non-Goals
++ Support automated **Bucket Creation**
++ Support automated **Bucket Deletion**
++ Support automated **Access Credential Generation**
```

Comment:
Let's say that I want to revoke the permissions of a user that abusing the system, correct me if I'm wrong, one issue with secrets is that you can't revoke them (until they expire?) once the user has them (you could delete the BAR and it could delete the secret but by that time they were exposed and the user might have copied them), what would happen in this case?

Reply:
One nice thing about IAM Roles is that you can create one such that you have access to a part of a bucket but not the entire bucket e.g.

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": "*",
      "Action": "s3:*",
      "Resource": [
        "arn:aws:s3:::MyExampleBucket/my/private/path"
      ]
    }
  ]
}
```

Is this possible to do with AWS Access Tokens?

Reply:
> It looks like we're not taking advantage of Service Accounts which already have the projection as a volume feature in kubernetes. I believe IAM Roles were specifically created so that we don't store credentials, it's interesting that I found info about accessing the temporary credentials in this place though https://aws.amazon.com/blogs/containers/cross-account-iam-roles-for-kubernetes-service-accounts/. Was it considered for this project?

Yes, we considered it for COSI. The issue with ServiceAccounts is an additional controller needs to be in place that maps pods to IAM credentials for that OSP. Non-cloudprovider OSP vendors do not support this universally.

> I read that credentials are going to be created on the fly, as a user running a workload in a pod is my understanding correct that a kubernetes Secret object is going to be created and possible made available to the workload [by making the secret a projected volume]? I think it'd be great to clarify how would a user be able to issue operations against the buckets too, for example once the user has the secret how would he use it?

Yes, it is explained in the "Bucket Provisioning for Pods" section.

> Let's say that I want to revoke the permissions of a user that abusing the system, correct me if I'm wrong, one issue with secrets is that you can't revoke them (until they expire?) once the user has them (you could delete the BAR and it could delete the secret but by that time they were exposed and the user might have copied them), what would happen in this case?

The IAM credentials can be revoked. The Kubernetes secrets will continue to exists, but will not be a valid access credential.

> A follow up question for the secret rotation, this is going to be taken care by the controller?

Well, it can be done by the controller, but there is no way to make the application client use the newly rotated credentials.

> shouldn't it be in a Pending state (not ready) while the bucket is being provisioned? If so then COSI might need to modify intree code to add a hook here, I'm not sure if this is what happens with File & Block Volumes. (NOTE: I just read the Bucket Provisioning part for the pods so it looks like this is what's going to happen)

Plain old secrets provide the same semantics. As in, the pod won't start until the secret exists.

> One nice thing about IAM Roles is that you can create one such that you have access to a part of a bucket but not the entire bucket e.g.

We support this use case, The properties field of the BAC can be used to specify this granular level of access control. That is a contract between the application developer and the driver however.

Reply:
> > It looks like we're not taking advantage of Service Accounts which already have the projection as a volume feature in kubernetes. I believe IAM Roles were specifically created so that we don't store credentials, it's interesting that I found info about accessing the temporary credentials in this place though https://aws.amazon.com/blogs/containers/cross-account-iam-roles-for-kubernetes-service-accounts/. Was it considered for this project?
> 
> Yes, we considered it for COSI. The issue with ServiceAccounts is an additional controller needs to be in place that maps pods to IAM credentials for that OSP. Non-cloudprovider OSP vendors do not support this universally.
> 
> > I read that credentials are going to be created on the fly, as a user running a workload in a pod is my understanding correct that a kubernetes Secret object is going to be created and possible made available to the workload [by making the secret a projected volume]? I think it'd be great to clarify how would a user be able to issue operations against the buckets too, for example once the user has the secret how would he use it?
> 
> Yes, it is explained in the "Bucket Provisioning for Pods" section.
> 
> > Let's say that I want to revoke the permissions of a user that abusing the system, correct me if I'm wrong, one issue with secrets is that you can't revoke them (until they expire?) once the user has them (you could delete the BAR and it could delete the secret but by that time they were exposed and the user might have copied them), what would happen in this case?
> 
> The IAM credentials can be revoked. The Kubernetes secrets will continue to exists, but will not be a valid access credential.

Is this handled as a separate API for the COSI-Driver or ProvisionerRevokeBucketAccess() will be used ?

> 
> > A follow up question for the secret rotation, this is going to be taken care by the controller?
> 
> Well, it can be done by the controller, but there is no way to make the application client use the newly rotated credentials.
> 
> > shouldn't it be in a Pending state (not ready) while the bucket is being provisioned? If so then COSI might need to modify intree code to add a hook here, I'm not sure if this is what happens with File & Block Volumes. (NOTE: I just read the Bucket Provisioning part for the pods so it looks like this is what's going to happen)
> 
> Plain old secrets provide the same semantics. As in, the pod won't start until the secret exists.
> 
> > One nice thing about IAM Roles is that you can create one such that you have access to a part of a bucket but not the entire bucket e.g.
> 
> We support this use case, The properties field of the BAC can be used to specify this granular level of access control. That is a contract between the application developer and the driver however.

Reply:
IAM style authentication is supported by this KEP now!